core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -421,7 +423,32 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     // the callback for sending a fetch response
     def sendResponseCallback(responsePartitionData: Map[TopicAndPartition, FetchResponsePartitionData]) {
-      val mergedResponseStatus = responsePartitionData ++ unauthorizedResponseStatus
+
+      val convertedResponseStatus =
+        // Need to down-convert message when consumer only takes magic value 0.
+        if (fetchRequest.versionId <= 1) {
+          responsePartitionData.map({ case (tp, data) =>
+            tp -> {
+              // We only do down-conversion when:
+              // 1. The message format version configured for the topic is using magic value > 0, and
+              // 2. The message set contains message whose magic > 0
+              // This is to reduce the message format conversion as much as possible. The conversion will only occur
+              // when new message format is used for the topic and we see an old request.
+              // Please notice that if the message format is changed from a higher version back to lower version this
+              // test might break because some messages in new message format can be delivered to consumer without
+              // format down conversion.
+              if (replicaManager.getMessageFormatVersion(tp).exists(_ > 0) &&
```

Comment:
Hi Jun, the tests on the configuration here is trying to address the following problem. 

After people just upgrade to 0.10.0.0, most of the fetch request are still in v1. In this case, if we check whether all the messages are v0 or not, we are essentially iterating over the file message set. By checking the configuration, we can avoid that.

After people set message format to v1, we will only do the iterative check for old consumers (hopefully there won't be many at that point). 

The caveat of this approach is mentioned in the comments. After user set the message format to v1, if they decide to change the message format back to v0. The old consumer may see message v1 because of the discrepancy between actual message format and the config as you pointed out.

Do you prefer to simply take the performance cost right after people finish upgrade? This cost will go away after all the clients are upgraded, but that could be an extended period.